# Draft analysis 

---

Group name: Group C, Burcin Aksel & Nico Henzel

---


## Introduction

### Inhalte aus Project Proposal übernehmen und konsolidieren
### Burcin

*This section includes an introduction to the project motivation, data, and research question. Include a data dictionary* 

## Setup

### Anpassen, sobald Import benötigt wird
### Nico

In [1]:
import pandas as pd
import numpy as np
import altair as alt
import matplotlib as mlp
import xlrd # needed to read excel files. install xlrd module if import does not work!
import shutil # needed to copy files


## Data

## Nico

### Inhalte aus Data Correction und Project Proposal übernehmen

## Import data

In [2]:
# Declare variables
external_data = '..\\data\\external\\'
raw_data = '..\\data\\raw\\'

# File names
orig_income_file = 'Median_Household_Income_By_State_1990-2017.xls'
target_income_file = 'Median_Household_Income_By_State_1990-2017.csv'
orig_death_file = 'NCHS_-_Leading_Causes_of_Death__United_States.csv'


In [3]:
# Save external median income data as csv in folder 'raw'
# Read file
xls_household_file = pd.read_excel(external_data+orig_income_file)
# Save file
xls_household_file.to_csv(raw_data+target_income_file,index = None, header=True)
# Copy external leading cause of death data into folder 'raw'
shutil.copy(external_data+orig_death_file, raw_data+orig_death_file)


'..\\data\\raw\\NCHS_-_Leading_Causes_of_Death__United_States.csv'

In [17]:
# Declare both dataframes
df_income = pd.read_csv(raw_data+target_income_file)
df_death = pd.read_csv(raw_data+orig_death_file)

### Data structure

In [12]:
df_income.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 68 entries, 0 to 67
Data columns (total 32 columns):
 #   Column                                                                              Non-Null Count  Dtype  
---  ------                                                                              --------------  -----  
 0   Table 102.30. Median household income, by state: Selected years, 1990 through 2017  59 non-null     object 
 1   Unnamed: 1                                                                          54 non-null     object 
 2   Unnamed: 2                                                                          54 non-null     object 
 3   Unnamed: 3                                                                          54 non-null     float64
 4   Unnamed: 4                                                                          52 non-null     float64
 5   Unnamed: 5                                                                          54 non-null     f

In [13]:
df_income

Table 102.30. Median household income, by state: Selected years, 1990 through 2017  \
0   [In constant 2017 dollars. Standard errors app...                                   
1                                               State                                   
2                                                   1                                   
3                           United States ...........                                   
4                        Alabama ....................                                   
..                                                ...                                   
63                         Wyoming ..................                                   
64  \1\Based on 1989 incomes collected in the 1990...                                   
65  \2\Based on 1999 incomes collected in the 2000...                                   
66  NOTE: Constant dollars adjusted by the Consume...                                   
67  SOURCE: U.S. Department of Commerce, Census Bu...                                   

   Unnamed: 1 Unnamed: 2  Unnamed: 3  Unnamed: 4  Unnamed: 5  Unnamed: 6  \
0         NaN        NaN         NaN         NaN         NaN         NaN   
1     1990\1\    2000\2\      2005.0         NaN      2010.0         NaN   
2           2          3         4.0         NaN         5.0         NaN   
3       57500      62000     58200.0        80.0     56400.0        40.0   
4       45200      50400     46400.0       400.0     45600.0       320.0   
..        ...        ...         ...         ...         ...         ...   
63      51900      55900     58100.0      1160.0     60300.0      1300.0   
64        NaN        NaN         NaN         NaN         NaN         NaN   
65        NaN        NaN         NaN         NaN         NaN         NaN   
66        NaN        NaN         NaN         NaN         NaN         NaN   
67        NaN        NaN         NaN         NaN         NaN         NaN   

    Unnamed: 7  Unnamed: 8  Unnamed: 9  ...  Unnamed: 22  Unnamed: 23  \
0          NaN         NaN         NaN  ...          NaN          NaN   
1       2013.0         NaN      2014.0  ...          NaN          NaN   
2          6.0         NaN         7.0  ...          NaN          NaN   
3      55100.0        40.0     55600.0  ...          NaN          NaN   
4      45200.0       410.0     44400.0  ...          NaN          NaN   
..         ...         ...         ...  ...          ...          ...   
63     61900.0      1150.0     59100.0  ...          NaN          NaN   
64         NaN         NaN         NaN  ...          NaN          NaN   
65         NaN         NaN         NaN  ...          NaN          NaN   
66         NaN         NaN         NaN  ...          NaN          NaN   
67         NaN         NaN         NaN  ...          NaN          NaN   

    Unnamed: 24  Unnamed: 25  Unnamed: 26  Unnamed: 27  Unnamed: 28  \
0           NaN          NaN          NaN          NaN          NaN   
1           NaN          NaN          NaN          NaN          NaN   
2           NaN          NaN          NaN          NaN          NaN   
3           NaN          NaN          NaN          NaN          NaN   
4           NaN          NaN          NaN          NaN          NaN   
..          ...          ...          ...          ...          ...   
63          NaN          NaN          NaN          NaN          NaN   
64          NaN          NaN          NaN          NaN          NaN   
65          NaN          NaN          NaN          NaN          NaN   
66          NaN          NaN          NaN          NaN          NaN   
67          NaN          NaN          NaN          NaN          NaN   

    Unnamed: 29  Unnamed: 30  Unnamed: 31  
0           NaN          NaN          NaN  
1           NaN          NaN          NaN  
2           NaN          NaN               
3           NaN          NaN          NaN  
4           NaN          NaN          NaN  
..          ...          ..

In [14]:
df_death.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10868 entries, 0 to 10867
Data columns (total 6 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Year                     10868 non-null  int64  
 1   113 Cause Name           10868 non-null  object 
 2   Cause Name               10868 non-null  object 
 3   State                    10868 non-null  object 
 4   Deaths                   10868 non-null  int64  
 5   Age-adjusted Death Rate  10868 non-null  float64
dtypes: float64(1), int64(2), object(3)
memory usage: 509.6+ KB


In [15]:
df_death

Year                                     113 Cause Name  \
0      2017  Accidents (unintentional injuries) (V01-X59,Y8...   
1      2017  Accidents (unintentional injuries) (V01-X59,Y8...   
2      2017  Accidents (unintentional injuries) (V01-X59,Y8...   
3      2017  Accidents (unintentional injuries) (V01-X59,Y8...   
4      2017  Accidents (unintentional injuries) (V01-X59,Y8...   
...     ...                                                ...   
10863  1999  Nephritis, nephrotic syndrome and nephrosis (N...   
10864  1999  Nephritis, nephrotic syndrome and nephrosis (N...   
10865  1999  Nephritis, nephrotic syndrome and nephrosis (N...   
10866  1999  Nephritis, nephrotic syndrome and nephrosis (N...   
10867  1999  Nephritis, nephrotic syndrome and nephrosis (N...   

                   Cause Name          State  Deaths  Age-adjusted Death Rate  
0      Unintentional injuries  United States  169936                     49.4  
1      Unintentional injuries        Alabama    2703                     53.8  
2      Unintentional injuries         Alaska     436                     63.7  
3      Unintentional injuries        Arizona    4184                     56.2  
4      Unintentional injuries       Arkansas    1625                     51.8  
...                       ...            ...     ...                      ...  
10863          Kidney disease       Virginia    1035                     16.9  
10864          Kidney disease     Washington     278                      5.2  
10865          Kidney disease  West Virginia     345                     16.4  
10866          Kidney disease      Wisconsin     677                     11.9  
10867          Kidney disease        Wyoming      30                      6.8  

[10868 rows x 6 columns]

In the death dataset we have 10868 cases and 6 columns.

The columns are:
* years, from 1999 to 2017
* 113 cause name, the NDI ICD-10 113 categories for causes of death
* cause name, the generic name for the death cause defined in the 113 cause name column
* state, in which state the data was collected
* death, the count of the total deaths 
* age-adjusted death rate, the standardized death percentage for the specific state in the observed year.


The following steps are:

* The income dataset needs to be cleaned and corrected in order to be joined with the death dataset (see data-corrections)
* The columns for the income dataset need to be transformed in order to have the same format as the death dataset.
* The join can be done by state, the missing income values for the years not represented in the income dataset need to either be dropped or replaced


### Data corrections

#### Income Dataset

In [42]:
df_income.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 68 entries, 0 to 67
Data columns (total 32 columns):
 #   Column                                                                              Non-Null Count  Dtype  
---  ------                                                                              --------------  -----  
 0   Table 102.30. Median household income, by state: Selected years, 1990 through 2017  59 non-null     object 
 1   Unnamed: 1                                                                          54 non-null     object 
 2   Unnamed: 2                                                                          54 non-null     object 
 3   Unnamed: 3                                                                          54 non-null     float64
 4   Unnamed: 4                                                                          52 non-null     float64
 5   Unnamed: 5                                                                          54 non-null     f

We will perform different corrections for this dataset. The steps are explained as comments within the code

In [161]:
# We only need the first 16 columns 
# and we can also drop the columns 4,6,8,10,12 and 14 since they only show NaN values
column_lst = [0,1,2,3,5,7,9,11,13,15]
df_income_corrected = df_income[df_income.columns[column_lst]]
# We don't need row 0,2 and row 64-67
row_drop_lst = [0,2,64,65,66,67]
df_income_corrected = df_income_corrected.drop(row_drop_lst)
# There are a few rows with empty entries left, which we can get rid off 
# since those rows have no state assigned to it, they are only used as separators
df_income_corrected.dropna(inplace=True)
# The column names are actually in the first row, additionally they need to be adjusted
column_names = ['State','1990','2000','2005','2010','2013','2014','2015','2016','2017']
df_income_corrected.columns = column_names
# Row number 1 can be dropped
df_income_corrected.drop(1,inplace=True)
# The dots in the state column can be removed
# We also do not want any leading or ending spaces in the strings
df_income_corrected = df_income_corrected.replace(r'\.','',regex=True)
df_income_corrected['State'] = df_income_corrected['State'].str.strip()
# Lastly we reset the row index drop the old index 
df_income_corrected.reset_index(drop = True, inplace = True)


We also need to make some data transformations in order for this dataset to be joined later on.

In [164]:
# We transform the table to show the median income for a state in a single year
# We use the melt() function for this
lst_years = column_names[1:]
df_income_corrected = df_income_corrected.melt(
    id_vars= ['State'], 
    value_vars= lst_years, 
    var_name= 'year', 
    value_name= 'median_household_income' 
    )

# The State column should be lowercase
df_income_corrected = df_income_corrected.rename(
    columns = {'State':'state'}
)
# The types need to be declared, state holds categorial values, year has integers and income holds float numbers
df_income_corrected['state'] = df_income_corrected['state'].astype('category')
df_income_corrected['year'] = df_income_corrected['year'].astype('int')
df_income_corrected['median_household_income'] = df_income_corrected['median_household_income'].astype('float')

In [165]:
df_income_corrected.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 468 entries, 0 to 467
Data columns (total 3 columns):
 #   Column                   Non-Null Count  Dtype   
---  ------                   --------------  -----   
 0   state                    468 non-null    category
 1   year                     468 non-null    int32   
 2   median_household_income  468 non-null    float64 
dtypes: category(1), float64(1), int32(1)
memory usage: 8.5 KB


In [166]:
df_income_corrected

state  year  median_household_income
0    United States  1990                  57500.0
1          Alabama  1990                  45200.0
2           Alaska  1990                  79300.0
3          Arizona  1990                  52700.0
4         Arkansas  1990                  40500.0
..             ...   ...                      ...
463       Virginia  2017                  71500.0
464     Washington  2017                  71000.0
465  West Virginia  2017                  43500.0
466      Wisconsin  2017                  59300.0
467        Wyoming  2017                  60400.0

[468 rows x 3 columns]

In [167]:
file = 'Corrected_Median_Household_Income_By_State_1990-2017.csv'
df_income_corrected.to_csv(raw_data+file,index = None, header=True)

We prepared the income data set in order to use it for our project. 
The original .xls data can be found in the data folder under external or the .csv under raw 

In the final death data set we have 468 cases and 3 columns.

The columns are:
* state, in which the data was collected
* years, 1990, 2000, 2005, 2010 and 2013-2017
* median_household_income

#### Death Dataset

In [168]:
df_death.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10868 entries, 0 to 10867
Data columns (total 6 columns):
 #   Column                   Non-Null Count  Dtype   
---  ------                   --------------  -----   
 0   year                     10868 non-null  int64   
 1   113_cause_name           10868 non-null  category
 2   cause_name               10868 non-null  category
 3   state                    10868 non-null  category
 4   deaths                   10868 non-null  int64   
 5   age_adjusted_death_rate  10868 non-null  float64 
dtypes: category(3), float64(1), int64(2)
memory usage: 289.9 KB


* The Dtypes 113 Cause Name, Cause Name and State need to be changed to category.
* The year, deaths and death rate column already have the right typing.
* The column names will be adjusted to be lowercase and have underscores instead of spaces.
* Also, there are no missing values present in the dataset.

In [178]:
df_death_corrected = df_death

In [179]:
# Change column names to lowercase
df_death_corrected.columns = df_death_corrected.columns.str.lower()
# Change spaces and the '-' to underscores
df_death_corrected.columns = df_death_corrected.columns.str.replace(r' ','_',regex=True)
df_death_corrected.columns = df_death_corrected.columns.str.replace(r'-','_',regex=True)
# Remove any leading or trailing whitespaces for the object columns
df_death_corrected['113_cause_name'] = df_death_corrected['113_cause_name'].str.strip()
df_death_corrected['cause_name'] = df_death_corrected['cause_name'].str.strip()
df_death_corrected['state'] = df_death_corrected['state'].str.strip()
# Change Dtype of 113 Cause Name, Cause Name and State column to category
cols = df_death_corrected.select_dtypes(include='object').columns.to_list()
df_death_corrected[cols] = df_death_corrected[cols].astype('category')

In [180]:
df_death_corrected.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10868 entries, 0 to 10867
Data columns (total 6 columns):
 #   Column                   Non-Null Count  Dtype   
---  ------                   --------------  -----   
 0   year                     10868 non-null  int64   
 1   113_cause_name           10868 non-null  category
 2   cause_name               10868 non-null  category
 3   state                    10868 non-null  category
 4   deaths                   10868 non-null  int64   
 5   age_adjusted_death_rate  10868 non-null  float64 
dtypes: category(3), float64(1), int64(2)
memory usage: 289.9 KB


In [181]:
df_joined = pd.merge(
    df_death_corrected,
    df_income_corrected,
    how = 'left',
    on = ['year','state']
)


In [184]:
df_joined.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10868 entries, 0 to 10867
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype   
---  ------                   --------------  -----   
 0   year                     10868 non-null  int64   
 1   113_cause_name           10868 non-null  category
 2   cause_name               10868 non-null  category
 3   state                    10868 non-null  category
 4   deaths                   10868 non-null  int64   
 5   age_adjusted_death_rate  10868 non-null  float64 
 6   median_household_income  4576 non-null   float64 
dtypes: category(3), float64(2), int64(2)
memory usage: 459.6 KB


After joining, the median_household_income is an additional column for the death data set.

We only have the median household income for the year 1990, 2000, 2005, 2010 and 2013-2017.
For the missing years, we will impute the values by using the last and next value and calculating the mean of both.


In [189]:
# Impute the median_household_income for missing years

This will leave us with 6 columns in total, which we can use for our data analysis (we will later disregard the 113 Cause name column since we will see, that the same information is provided by the cause name column)

### Variable lists
## Burcin

In [ ]:
# auf Fertigem Datensatz Variablen auflisten

### Data splitting

## Burcin

## Analysis

## Nico

Analysis Approach umsetzen

### Descriptive statistics

### Exploratory data analysis

### Relationships

## Model

## Nico
### Kopieren aus Notebook 44 und Modeltypen passend zu proposal umsetzen

Abstimmung, welche Metriken und Erkenntnisse aus den Modellen für uns relevant sind

### Select model

### Training and validation

### Fit model

### Evaluation on test set

### Save model



Save your model in the folder `models/`. Use a meaningful name and a timestamp.

## Conclusions

Am Ende die weiteren Schritte und bisherigen Interprationen auflisten.